pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		panic(err)
+	}
+
+	return &RemoteConfigHandler{
+		remoteClient:          conf.RemoteSamplingClient,
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		if configID == dynamicRatesByService || configID == userDefinedRatesByService {
+			var payload apmsampling.APMSampling
+			_, err := payload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal APMSampling remote config envPayload: %w", err)
+			}
+			remoteRateUpdates = append(remoteRateUpdates, sampler.RemoteRateUpdate{Version: conf.Metadata.Version, Config: payload})
+		}
+		if configID == samplerconfig {
+			samplerconfigPayload = &apmsampling.SamplerConfig{}
+			_, err := samplerconfigPayload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal SamplerConfig remote config envPayload: %w", err)
+			}
+		}
+	}
+	return remoteRateUpdates, samplerconfigPayload, nil
+}
+
+func (h *RemoteConfigHandler) configPathToID(path string) (apmsamplingRemoteConfigID, error) {
+	pathMatch := h.remoteConfigPathRegex.FindStringSubmatch(path)
+	if pathMatch == nil {
+		return "", fmt.Errorf("failed to match remote config path %s", path)
+	}
+
+	switch pathMatch[1] {
+	case string(dynamicRatesByService):
+		return dynamicRatesByService, nil
+	case string(userDefinedRatesByService):
+		return userDefinedRatesByService, nil
+	case string(samplerconfig):
+		return samplerconfig, nil
+	default:
+		return "", fmt.Errorf("unexpected remote config ID: %s", pathMatch[1])
+	}
+}
+
+func (h *RemoteConfigHandler) updateSamplers(config apmsampling.SamplerConfig) {
+	var confForEnv *apmsampling.SamplerEnvConfig
+	for _, envAndConfig := range config.ByEnv {
+		if envAndConfig.Env == h.agentConfig.DefaultEnv {
```

Comment:
Am I understanding this correctly? We ignore any `ByEnv` configuration that isn't for the `DefaultEnv`?

Reply:
This is correct.

`DefaultEnv` is the agent's environment, which is applied by default to traces if the env is not specified by the tracer (IIUC).

So the rationale here is that the user can choose to configure agents in certains environments only. This configuration "by environment" should take precedence over the more general "for all environments" config.

Hope this makes sense :)